[SaferCPP] Fix some lambda captures in WebCore accessibility

atar13 opened the pull request:
#### f5039282ca9e8eab0b07dcba6a5b8d0c217053d6
<pre>
[SaferCPP] Fix some lambda captures in WebCore accessibility
<a href="https://bugs.webkit.org/show_bug.cgi?id=299993">https://bugs.webkit.org/show_bug.cgi?id=299993</a>
<a href="https://rdar.apple.com/161779222">rdar://161779222</a>

Reviewed by Tyler Wilcock.

Lambda capture warnings such as &quot;Implicitly captured raw-pointer &apos;this&apos; to uncounted
type is unsafe&quot; and &quot;Captured raw-pointer &apos;this&apos; to uncounted type is unsafe&quot; in
AccessibilityScrollView.cpp and AXIsolatedTree.cpp.

Captures this after making it Ref counted in capture initialization.

* Source/WebCore/SaferCPPExpectations/UncountedLambdaCapturesCheckerExpectations:
* Source/WebCore/accessibility/AccessibilityScrollView.cpp:
(WebCore::AccessibilityScrollView::addRemoteFrameChild):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.cpp:
(WebCore::AXIsolatedTree::updatePropertiesForSelfAndDescendants):
(WebCore::AXIsolatedTree::applyPendingChangesLocked):

Canonical link: <a href="https://commits.webkit.org/300867@main">https://commits.webkit.org/300867@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d4096d76a9b9c69d9bf004c197bc28a75de80247

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/124150 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/43843 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/34560 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/130977 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/76251 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/126027 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/44584 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/52443 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/94435 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/62648 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/127104 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/35521 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/111048 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/75026 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/34467 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/29209 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/74463 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/105262 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/29430 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/133651 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/51075 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/38920 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/102908 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/51459 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/107265 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/102715 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/26126 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/48066 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/26315 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/47953 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/50934 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/56709 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/50384 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/53736 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/52061 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/791573fb2772d2fa44e7c8b7fae8c8e9f43db03c. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51659

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/8948cb5510fb6440c87eb301e0df6f0e75090f11. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51659

atar13 commented on Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.cpp:
```diff
@@ -1395,7 +1395,7 @@ void AXIsolatedTree::applyPendingChangesLocked()
         // analysis throws a false-positive compile error when we access m_pendingProtectedFromDeletionIDs in
         // this lambda.
-        std::function<void(Ref<AXCoreObject>&&)> deleteSubtree = [&] (Ref<AXCoreObject>&& coreObjectToDelete) WTF_IGNORES_THREAD_SAFETY_ANALYSIS {
+        std::function<void(Ref<AXCoreObject>&&)> deleteSubtree = [this, protectedThis = Ref { *this }, &deleteSubtree] (Ref<AXCoreObject>&& coreObjectToDelete) WTF_IGNORES_THREAD_SAFETY_ANALYSIS {
```
I wanted to make sure this case was ok since this lambda `deleteSubtree` recursively calls itself. The ref count on `this` should only be incremented once when the lambda is defined right?

twilco replied:
Right, we only bump the ref count once when the lambda is created.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/d4096d76a9b9c69d9bf004c197bc28a75de80247. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51659

atar13 committed:
[SaferCPP] Fix some lambda captures in WebCore accessibility
https://bugs.webkit.org/show_bug.cgi?id=299993
rdar://161779222

Reviewed by Tyler Wilcock.

Lambda capture warnings such as "Implicitly captured raw-pointer 'this' to uncounted
type is unsafe" and "Captured raw-pointer 'this' to uncounted type is unsafe" in
AccessibilityScrollView.cpp and AXIsolatedTree.cpp.

Captures this after making it Ref counted in capture initialization.

* Source/WebCore/SaferCPPExpectations/UncountedLambdaCapturesCheckerExpectations:
* Source/WebCore/accessibility/AccessibilityScrollView.cpp:
(WebCore::AccessibilityScrollView::addRemoteFrameChild):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.cpp:
(WebCore::AXIsolatedTree::updatePropertiesForSelfAndDescendants):
(WebCore::AXIsolatedTree::applyPendingChangesLocked):

Canonical link: https://commits.webkit.org/300867@main
